File: terceiro_semestre/tecnicas/LEE04_py/ex002/filhos.py
def filho_esquerdo(n):
    return 2 * n

def filho_direito(n):
    return 2 * n + 1

def main():
    n, c = map(int, input().split())

    arvore = [-1] * (n + 1)
    valores = list(map(int, input().split()))
    for i in range(1, n + 1):
        arvore[i] = valores[i - 1]

    for _ in range(c):
        no_consulta = int(input())

        if arvore[1] == -1:
            print("NULL")
            break

        esquerdo = filho_esquerdo(no_consulta)
        direito = filho_direito(no_consulta)

        if esquerdo <= n and arvore[esquerdo] != -1:
            print(arvore[esquerdo], end=" ")
        else:
            print("NULL", end=" ")

        if direito <= n and arvore[direito] != -1:
            print(arvore[direito])
        else:
            print("NULL")

if __name__ == "__main__":
    main()

File: terceiro_semestre/tecnicas/LEE04/ex002/filhos.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

int main() {
    int N, C;
    cin >> N >> C;

    vector<int> valores(N);
    for (int i = 0; i < N; i++) {
        cin >> valores[i];
    }

    vector<int> consultas(C);
    for (int i = 0; i < C; i++) {
        cin >> consultas[i];
        consultas[i]--;  
    }

    for (int consulta : consultas) {
        int esquerdo = 2 * consulta + 1;
        int direito = 2 * consulta + 2;

        string filho_esquerdo = (esquerdo < N && valores[esquerdo] != -1) ? to_string(valores[esquerdo]) : "NULL";
        string filho_direito = (direito < N && valores[direito] != -1) ? to_string(valores[direito]) : "NULL";

        cout << filho_esquerdo << " " << filho_direito << endl;
    }

    return 0;
}

// =============================================================================================================================

#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<string> find_children(const vector<int>& tree, int N, const vector<int>& queries) {
    vector<string> results;
    
    for (int node : queries) {
        string left_child = "NULL";
        string right_child = "NULL";
        
        if (node - 1 >= 0 && node - 1 < N && tree[node - 1] != -1) {
            for (int current_pos = 0; current_pos < N; current_pos++) {
                if (tree[current_pos] == tree[node - 1]) {
                    if (current_pos * 2 + 1 < N && tree[current_pos * 2 + 1] != -1) {
                        left_child = to_string(tree[current_pos * 2 + 1]);
                    }
                    if (current_pos * 2 + 2 < N && tree[current_pos * 2 + 2] != -1) {
                        right_child = to_string(tree[current_pos * 2 + 2]);
                    }
                    break;
                }
            }
        }
        
        results.push_back(left_child + " " + right_child);
    }
    
    return results;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int N, C;
    cin >> N >> C;
    
    vector<int> tree(N);
    for (int i = 0; i < N; i++) {
        cin >> tree[i];
    }
    
    vector<int> queries(C);
    for (int i = 0; i < C; i++) {
        cin >> queries[i];
    }
    
    vector<string> results = find_children(tree, N, queries);

    for (const string& result : results) {
        cout << result << endl;
    }
    
    return 0;
}